app/jobs/fetch_hearing_locations_for_veterans_job.rb

```
@@ -55,34 +61,63 @@ def perform
+      .pluck(:facility_locator_id)
   end
 
-  def facility_ids_for_veteran(veteran)
-    file_number = LegacyAppeal.convert_file_number_to_vacols(veteran.file_number)
-    ro = ros_hash[file_number]
+  def fetch_and_update_ro_for_veteran(veteran, va_dot_gov_address:)
+    state_code = get_state_code(va_dot_gov_address)
+    facility_ids = ro_facility_ids_for_state(state_code)
 
-    RegionalOffice::CITIES[ro][:alternate_locations] || [] << RegionalOffice::CITIES[ro][:facility_locator_id]
+    distances = VADotGovService.get_distance(
+      lat: va_dot_gov_address[:lat], long: va_dot_gov_address[:long], ids: facility_ids
+    )
+
+    closest_ro_index = RegionalOffice::CITIES.values.find_index { |ro| ro[:facility_locator_id] == distances[0][:id] }
+    closest_ro = RegionalOffice::CITIES.keys[closest_ro_index]
+    veteran.update(closest_regional_office: closest_ro)
+
+    closest_ro
   end
 
-  def full_address_for(address)
-    address_1 = address["address_1"]
-    address_2 = address["address_2"].blank? ? "" : " " + address["address_2"]
-    address_3 = address["address_3"].blank? ? "" : " " + address["address_3"]
+  def valid_states
+    @valid_states ||= RegionalOffice::CITIES.values.reject { |ro| ro[:facility_locator_id].nil? }.pluck(:state)
+  end
 
-    "#{address_1}#{address_2}#{address_3}"
+  def get_state_code(va_dot_gov_address)
+    state_code = case va_dot_gov_address[:country_code]
+                 # Guam, American Somao, Marshall Islands, Micronesia, Northern Mariana Islands, Palau
+                 when "GQ", "AQ", "RM", "FM", "CQ", "PS"
+                   "HI"
+                 when "PH", "RP", "PI"
+                   "PI"
+                 when "VI", "VQ", "PR"
+                   "PR"
+                 when "US"
+                   va_dot_gov_address[:state_code]
+                 else
+                   msg = "#{va_dot_gov_address[:country_code]} is not a valid country code."
```

We might need a better story for Vets outside the country not in these places.

---

But let's see when we error!

---

@MeredithStewart and I discussed this briefly. I think we agreed it would be good to create block for the ScheduleVeteranTask if the Vet's address is outside US territories. @sharonwarner should we add this to a sprint? We might be able to get away with just failing silently if we can't fit this in before Feb 14th.

---

created issue #8614